File: sem3l3go/dlinkedlist/dlinkedlist.go
package dlinkedlist

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

// DlistNode представляет узел двусвязного списка
type DlistNode struct {
	Data string
	Next *DlistNode
	Prev *DlistNode
}

// DlinkedList представляет двусвязный список
type DlinkedList struct {
	Head *DlistNode
	Tail *DlistNode
}

// NewDlinkedList создает новый пустой список
func NewDlinkedList() *DlinkedList {
	return &DlinkedList{
		Head: nil,
		Tail: nil,
	}
}

// AddToHead добавляет элемент в начало
func (dll *DlinkedList) AddToHead(value string) {
	newNode := &DlistNode{Data: value, Next: dll.Head, Prev: nil}
	if dll.Head != nil {
		dll.Head.Prev = newNode
	}
	dll.Head = newNode
	if dll.Tail == nil {
		dll.Tail = dll.Head
	}
}

// AddToTail добавляет элемент в конец
func (dll *DlinkedList) AddToTail(value string) {
	newNode := &DlistNode{Data: value, Next: nil, Prev: dll.Tail}
	if dll.Tail != nil {
		dll.Tail.Next = newNode
	}
	dll.Tail = newNode
	if dll.Head == nil {
		dll.Head = dll.Tail
	}
}

// RemoveFromHead удаляет элемент с головы
func (dll *DlinkedList) RemoveFromHead() bool {
	if dll.Head == nil {
		return false
	}
	dll.Head = dll.Head.Next
	if dll.Head != nil {
		dll.Head.Prev = nil
	} else {
		dll.Tail = nil
	}
	return true
}

// RemoveFromTail удаляет элемент с хвоста
func (dll *DlinkedList) RemoveFromTail() bool {
	if dll.Tail == nil {
		return false
	}
	dll.Tail = dll.Tail.Prev
	if dll.Tail != nil {
		dll.Tail.Next = nil
	} else {
		dll.Head = nil
	}
	return true
}

// RemoveByValue удаляет элемент по значению
func (dll *DlinkedList) RemoveByValue(value string) bool {
	for current := dll.Head; current != nil; current = current.Next {
		if current.Data == value {
			if current.Prev != nil {
				current.Prev.Next = current.Next
			} else {
				dll.Head = current.Next
			}
			if current.Next != nil {
				current.Next.Prev = current.Prev
			} else {
				dll.Tail = current.Prev
			}
			return true
		}
	}
	return false
}

// Search ищет элемент по значению
func (dll *DlinkedList) Search(value string) bool {
	for current := dll.Head; current != nil; current = current.Next {
		if current.Data == value {
			return true
		}
	}
	return false
}

// Print выводит все элементы
func (dll *DlinkedList) Print() {
	for current := dll.Head; current != nil; current = current.Next {
		fmt.Printf("%s ", current.Data)
	}
	fmt.Println()
}

// Destroy очищает список
func (dll *DlinkedList) Destroy() {
	for dll.Head != nil {
		dll.RemoveFromHead()
	}
}

// LoadFromFile загружает из файла
func (dll *DlinkedList) LoadFromFile(filename string) error {
	if filename == "" {
		return nil
	}
	file, err := os.Open(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		dll.AddToTail(scanner.Text())
	}
	return scanner.Err()
}

// SaveToFile сохраняет в файл
func (dll *DlinkedList) SaveToFile(filename string) error {
	if filename == "" {
		return nil
	}
	file, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	for current := dll.Head; current != nil; current = current.Next {
		if _, err := fmt.Fprintln(file, current.Data); err != nil {
			return err
		}
	}
	return nil
}

// AddBefore добавляет элемент перед target
func (dll *DlinkedList) AddBefore(target, value string) bool {
	if dll.Head == nil {
		return false
	}

	// Если target в голове
	if dll.Head.Data == target {
		dll.AddToHead(value)
		return true
	}

	// Ищем target
	current := dll.Head
	for current != nil && current.Data != target {
		current = current.Next
	}
	if current == nil {
		return false
	}

	// Вставляем перед current
	newNode := &DlistNode{Data: value, Next: current, Prev: current.Prev}
	if current.Prev != nil {
		current.Prev.Next = newNode
	}
	current.Prev = newNode
	return true
}

// AddAfter добавляет элемент после target
func (dll *DlinkedList) AddAfter(target, value string) bool {
	current := dll.Head
	for current != nil && current.Data != target {
		current = current.Next
	}
	if current == nil {
		return false
	}

	newNode := &DlistNode{Data: value, Next: current.Next, Prev: current}
	if current.Next != nil {
		current.Next.Prev = newNode
	} else {
		dll.Tail = newNode
	}
	current.Next = newNode
	return true
}

// RemoveBefore удаляет элемент перед target
func (dll *DlinkedList) RemoveBefore(target string) bool {
	if dll.Head == nil || dll.Head.Data == target {
		return false
	}

	// Если второй элемент - target
	if dll.Head.Next != nil && dll.Head.Next.Data == target {
		return dll.RemoveFromHead()
	}

	// Ищем target
	current := dll.Head
	for current != nil && current.Data != target {
		current = current.Next
	}
	if current == nil || current.Prev == nil {
		return false
	}

	// Удаляем элемент перед target
	toRemove := current.Prev
	if toRemove.Prev != nil {
		toRemove.Prev.Next = current
		current.Prev = toRemove.Prev
	} else {
		// Если удаляем первый элемент
		dll.Head = current
		current.Prev = nil
	}
	return true
}

// RemoveAfter удаляет элемент после target
// RemoveAfter удаляет элемент после target
func (dll *DlinkedList) RemoveAfter(target string) bool {
	if dll.Head == nil {
		return false
	}

	current := dll.Head
	for current != nil && current.Data != target {
		current = current.Next
	}
	if current == nil || current.Next == nil {
		return false
	}

	toRemove := current.Next
	current.Next = toRemove.Next
	if toRemove.Next != nil {
		toRemove.Next.Prev = current
	} else {
		dll.Tail = current
	}
	return true
}

// RunDLinkedList выполняет команды над списком
func RunDLinkedList(args []string) {
	list := NewDlinkedList()
	var filename string
	var query string

	for i := 0; i < len(args); i++ {
		if args[i] == "--file" && i+1 < len(args) {
			filename = args[i+1]
			i++
		} else if args[i] == "--query" && i+1 < len(args) {
			query = args[i+1]
			i++
		}
	}

	if filename != "" {
		list.LoadFromFile(filename)
	}

	parts := strings.SplitN(query, " ", 3)
	command := parts[0]
	token1, token2 := "", ""
	if len(parts) > 1 {
		token1 = parts[1]
	}
	if len(parts) > 2 {
		token2 = parts[2]
	}

	switch command {
	case "DPUSH":
		if token1 != "" {
			list.AddToHead(token1)
			if filename != "" {
				list.SaveToFile(filename)
			}
		}
	case "DAPPEND":
		if token1 != "" {
			list.AddToTail(token1)
			if filename != "" {
				list.SaveToFile(filename)
			}
		}
	case "DREMOVEHEAD":
		list.RemoveFromHead()
		if filename != "" {
			list.SaveToFile(filename)
		}
	case "DREMOVETAIL":
		list.RemoveFromTail()
		if filename != "" {
			list.SaveToFile(filename)
		}
	case "DREMOVE":
		if token1 != "" {
			list.RemoveByValue(token1)
			if filename != "" {
				list.SaveToFile(filename)
			}
		}
	case "DSEARCH":
		if token1 != "" {
			if list.Search(token1) {
				fmt.Println("true")
			} else {
				fmt.Println("false")
			}
		}
	case "DPRINT":
		list.Print()
	case "DADDTO":
		if token1 != "" && token2 != "" {
			list.AddBefore(token1, token2)
			if filename != "" {
				list.SaveToFile(filename)
			}
		}
	case "DADDAFTER":
		if token1 != "" && token2 != "" {
			list.AddAfter(token1, token2)
			if filename != "" {
				list.SaveToFile(filename)
			}
		}
	case "DREMOVETO":
		if token1 != "" {
			list.RemoveBefore(token1)
			if filename != "" {
				list.SaveToFile(filename)
			}
		}
	case "DREMOVEAFTER":
		if token1 != "" {
			list.RemoveAfter(token1)
			if filename != "" {
				list.SaveToFile(filename)
			}
		}
	}

	list.Destroy()
}

File: C++/Laba3sem2v2/dlinkedList.cpp
#include "dlinkedList.h"
#include <cstring>
#include <fstream>
#include <iostream>
#include <sstream>

using namespace std;

// Инициализация списка.
void DlinkedList::init() {
    head = nullptr;
    tail = nullptr;
}

// Добавление нового элемента в начало.
void DlinkedList::addToHead(const string& value) {
    DlistNode* newNode = new DlistNode{ value, head, nullptr };
    if (head != nullptr) {
        head->prev = newNode;
    }
    head = newNode;
    if (tail == nullptr) {
        tail = head;
    }
}

// Добавление нового элемента в конец.
void DlinkedList::addToTail(const string& value) {
    DlistNode* newNode = new DlistNode{ value, nullptr, tail };
    if (tail != nullptr) {
        tail->next = newNode;
    }
    tail = newNode;
    if (head == nullptr) {
        head = tail;
    }
}

// Удаление элемента с головы.
void DlinkedList::removeFromHead() {
    if (head == nullptr) {
        return;
    }
    DlistNode* temp = head;
    head = head->next;
    if (head != nullptr) {
        head->prev = nullptr;
    }
    else {
        tail = nullptr;
    }
    delete temp;
}

// Удаление элемента с хвоста.
void DlinkedList::removeFromTail() {
    if (tail == nullptr) {
        return;
    }
    DlistNode* temp = tail;
    tail = tail->prev;
    if (tail != nullptr) {
        tail->next = nullptr;
    }
    else {
        head = nullptr;
    }
    delete temp;
}

// Удаление узла по значению.
void DlinkedList::removeByValue(const string& value) {
    DlistNode* temp = head;
    while (temp != nullptr) {
        if (temp->data == value) {
            if (temp->prev != nullptr) {
                temp->prev->next = temp->next;
            }
            else {
                head = temp->next;
            }
            if (temp->next != nullptr) {
                temp->next->prev = temp->prev;
            }
            else {
                tail = temp->prev;
            }
            delete temp;
            return;
        }
        temp = temp->next;
    }
}

// Поиск элемента по значению.
bool DlinkedList::search(const string& value) {
    DlistNode* temp = head;
    while (temp != nullptr) {
        if (temp->data == value) {
            return true;
        }
        temp = temp->next;
    }
    return false;
}

// Вывод всех элементов списка.
void DlinkedList::print() {
    DlistNode* temp = head;
    while (temp != nullptr) {
        cout << temp->data << " ";
        temp = temp->next;
    }
    cout << endl;
}

// Очистка списка.
void DlinkedList::destroy() {
    while (head != nullptr) {
        removeFromHead();
    }
}

// Загрузка элементов из файла.
void DlinkedList::loadFromFile(const string& fileName) {
    if (fileName.empty()) return; // ДОБАВЛЕНА ПРОВЕРКА
    ifstream file(fileName);
    string value;
    while (file >> value) {
        addToTail(value);
    }
    file.close();
}

// Сохранение элементов списка в файл.
void DlinkedList::saveToFile(const string& fileName) {
    if (fileName.empty()) return; // ДОБАВЛЕНА ПРОВЕРКА
    ofstream file(fileName);
    DlistNode* temp = head;
    while (temp != nullptr) {
        file << temp->data << endl;
        temp = temp->next;
    }
    file.close();
}

// --------------------- Новые методы ---------------------

void DlinkedList::addBefore(const string& target, const string& value) {
    if (head == nullptr) return;

    // если target в голове — просто добавить в голову
    if (head->data == target) {
        addToHead(value);
        return;
    }

    DlistNode* node = head->next;
    while (node != nullptr && node->data != target) {
        node = node->next;
    }
    if (node == nullptr) return; // target не найден

    // вставляем перед node
    DlistNode* newNode = new DlistNode{ value, node, node->prev };
    if (node->prev != nullptr) node->prev->next = newNode;
    node->prev = newNode;
}

void DlinkedList::addAfter(const string& target, const string& value) {
    DlistNode* node = head;
    while (node != nullptr && node->data != target) {
        node = node->next;
    }
    if (node == nullptr) return; // target не найден

    DlistNode* newNode = new DlistNode{ value, node->next, node };
    if (node->next != nullptr) {
        node->next->prev = newNode;
    }
    else {
        // вставка после хвоста -> обновляем tail
        tail = newNode;
    }
    node->next = newNode;
}

void DlinkedList::removeBefore(const string& target) {
    if (head == nullptr) return;

    // если target находится в голове — ничего удалить
    if (head->data == target) return;

    // если второй элемент — удалить голову
    if (head->next != nullptr && head->next->data == target) {
        removeFromHead();
        return;
    }

    DlistNode* node = head->next;
    while (node != nullptr && node->data != target) {
        node = node->next;
    }
    if (node == nullptr) return; // target не найден

    DlistNode* toRemove = node->prev;
    if (toRemove == nullptr) return; // нет элемента перед target (безопасность)

    // удалить toRemove
    DlistNode* before = toRemove->prev;
    node->prev = before;
    if (before != nullptr) {
        before->next = node;
    }
    else {
        // удалён был head
        head = node;
    }
    delete toRemove;
}

void DlinkedList::removeAfter(const string& target) {
    DlistNode* node = head;
    while (node != nullptr && node->data != target) {
        node = node->next;
    }
    if (node == nullptr) return; // target не найден
    if (node->next == nullptr) return; // нет элемента после target

    DlistNode* toRemove = node->next;
    DlistNode* after = toRemove->next;
    node->next = after;
    if (after != nullptr) {
        after->prev = node;
    }
    else {
        // удалён был tail
        tail = node;
    }
    delete toRemove;
}

// ---------------------------------------------------------

// Функция запуска работы со списком, обработка команд из аргументов командной строки.
void runLLinkedList(int argc, char* argv[]) {
    DlinkedList list;
    list.init();

    string fileName;
    string query;

    for (int i = 1; i < argc; i++) {
        if (strcmp(argv[i], "--file") == 0 && i + 1 < argc) {
            fileName = argv[i + 1];
            i++;
        }
        else if (strcmp(argv[i], "--query") == 0 && i + 1 < argc) {
            query = argv[i + 1];
            i++;
        }
    }

    if (!fileName.empty())
        list.loadFromFile(fileName);

    string command;
    string rest;
    size_t pos = query.find(' ');
    if (pos != string::npos) {
        command = query.substr(0, pos);
        rest = query.substr(pos + 1);
    }
    else {
        command = query;
    }

    // разбиваем остальную часть запроса на токены
    istringstream iss(rest);
    string token1, token2;
    iss >> token1;
    iss >> token2; // может быть пустым

    if (command == "DPUSH") {
        if (!token1.empty()) {
            list.addToHead(token1);
            if (!fileName.empty()) list.saveToFile(fileName);
        }
    }
    else if (command == "DAPPEND") {
        if (!token1.empty()) {
            list.addToTail(token1);
            if (!fileName.empty()) list.saveToFile(fileName);
        }
    }
    else if (command == "DREMOVEHEAD") {
        list.removeFromHead();
        if (!fileName.empty()) list.saveToFile(fileName);
    }
    else if (command == "DREMOVETAIL") {
        list.removeFromTail();
        if (!fileName.empty()) list.saveToFile(fileName);
    }
    else if (command == "DREMOVE") {
        if (!token1.empty()) {
            list.removeByValue(token1);
            if (!fileName.empty()) list.saveToFile(fileName);
        }
    }
    else if (command == "DSEARCH") {
        if (!token1.empty())
            cout << (list.search(token1) ? "true" : "false") << endl;
    }
    else if (command == "DPRINT") {
        list.print();
    }
    // Новые команды:
    else if (command == "DADDTO") {
        // формат: DADDTO <target> <value>  -> вставить <value> до <target>
        if (!token1.empty() && !token2.empty()) {
            list.addBefore(token1, token2);
            if (!fileName.empty()) list.saveToFile(fileName);
        }
    }
    else if (command == "DADDAFTER") {
        // формат: DADDAFTER <target> <value>  -> вставить <value> после <target>
        if (!token1.empty() && !token2.empty()) {
            list.addAfter(token1, token2);
            if (!fileName.empty()) list.saveToFile(fileName);
        }
    }
    else if (command == "DREMOVETO") {
        // формат: DREMOVETO <target>  -> удалить элемент до <target>
        if (!token1.empty()) {
            list.removeBefore(token1);
            if (!fileName.empty()) list.saveToFile(fileName);
        }
    }
    else if (command == "DREMOVEAFTER") {
        // формат: DREMOVEAFTER <target> -> удалить элемент после <target>
        if (!token1.empty()) {
            list.removeAfter(token1);
            if (!fileName.empty()) list.saveToFile(fileName);
        }
    }

    list.destroy();
}